src/text_page.rs

```diff
@@ -52,33 +54,40 @@ impl TextPage {
     }
 
     pub fn search(&self, needle: &str, hit_max: u32) -> Result<Vec<Quad>, Error> {
-        struct Quads(*mut fz_quad);
-
-        impl Drop for Quads {
-            fn drop(&mut self) {
-                if !self.0.is_null() {
-                    unsafe { fz_free(context(), self.0 as _) };
-                }
-            }
-        }
-
         let c_needle = CString::new(needle)?;
         let hit_max = if hit_max < 1 { 16 } else { hit_max };
         let mut hit_count = 0;
-        unsafe {
-            let quads = Quads(ffi_try!(mupdf_search_stext_page(
+        let quads = unsafe {
+            ffi_try!(mupdf_search_stext_page(
                 context(),
                 self.inner,
                 c_needle.as_ptr(),
                 hit_max as _,
                 &mut hit_count
-            )));
-            if hit_count == 0 {
-                return Ok(Vec::new());
-            }
-            let items = slice::from_raw_parts(quads.0, hit_count as usize);
-            Ok(items.iter().map(|quad| (*quad).into()).collect())
-        }
+            ))
+        };
+
+        unsafe { rust_vec_from_ffi_ptr(quads, hit_count) }
+    }
+
+    pub fn highlight_selection(
+        &mut self,
+        a: Point,
+        b: Point,
+        quads: &[Quad],
+    ) -> Result<i32, Error> {
+        let (ptr, len): (*const fz_quad, _) = rust_slice_to_ffi_ptr(quads)?;
+
+        Ok(unsafe {
+            fz_highlight_selection(
```

Comment:
this should be wrapped in `ffi_try!` I think, for example they did it in Java binding: https://github.com/ArtifexSoftware/mupdf/blob/9ebfcf2465b858b8c0e2575c657833e9418550c6/platform/java/jni/structuredtext.c#L77-L80

Reply:
`ffi_try!` only works for functions with take an error pointer as the last argument - this function doesn't, so ffi_try won't work with it. It seems there's no way for it to error.

Reply:
That's why we have `wrapper.c` to wrap them with `fz_try` and `fz_catch`. While I agree that it seem there's no way for it to error at the moment, there is no guarantee that it won't start to throw errors in the future. Anyway I guess we can deal with that later when breakage actually happens, not a big deal.

Reply:
Oh, I was under the impression that `mupdf` promised to never throw, and instead just propogate errors via a `*mut *mut mupdf_error_t` passed in. I realize now that those error ptrs are actually added via the `wrapper.c` file adding them in.

So, to make sure I'm following the process for this library: `wrapper.c` is manually written, and we want to make sure that it is future-proofed by adding each bridged function to it and wrapping that function in `fz_try(ctx) {} fz_catch(ctx) {}...` blocks so that we can catch and return errors before they reach the FFI layer. Is that correct?

(I know you already resolved this, I would just like to get this clarified if possible, and fixed as well since it seems I misunderstood how `ffi_try` was used and implemented)

Reply:
Correct.